### PR TITLE
CORE-385:   Add TTL + automatic deletion of instances

### PR DESCRIPTION
**Pull request by cored**
Fix: #80

### CODE REVIEW

File: instance/instance.go [hunk 1]
```diff
@@ -4,10 +4,17 @@ import (
...
 	"errors"
 	"fmt"
+	"time"
 
 	"github.com/namely/broadway/store"
 )
 
+// NewExpiredAt builds a new ExpiredAt
+func NewExpiredAt(daysToExpire int, from time.Time) time.Time {
+	from.AddDate(from.Year(), int(from.Month()), from.Day()+daysToExpire)
```
Comment (therealplato) [inline instance/instance.go]: what happens when from.Day is 29 and you add 5 days to expire?
how about `return from.Add(time.Day * time.Duration(daysToExpire))`

File: instance/instance.go [hunk 2]
```diff
@@ -177,3 +187,17 @@ func toJSON(instance *Instance) (string, error) {
 	}
 	return string(encoded), nil
 }
+
+func retrieveInstancesByKey(store store.Store, key string) ([]*Instance, error) {
+	data := store.Values(key)
+	instances := []*Instance{}
```
Comment (therealplato) [inline instance/instance.go]: can you use `[]Instance` here?

File: instance/instance.go [hunk 3]
```diff
@@ -25,6 +32,13 @@ func (e NotFoundError) Error() string {
 // ErrMalformedSaveData in case an instance cannot be marshall
 var ErrMalformedSaveData = errors.New("broadway/instance: saved data for this instance is malformed")
 
+// GlobalPath represents the global path key to save an instance in store
+type GlobalPath string
+
+func (gp GlobalPath) String() string {
+	return fmt.Sprintf("%s/instances", string(gp))
+}
+
```
Comment (therealplato) [inline instance/instance.go]: i dont think this is any better than `var GlobalPath = PathPrefix + "/instances"`

File: services/instance.go [hunk 4]
```diff
@@ -15,6 +15,9 @@ import (
 	"github.com/namely/broadway/store"
 )
 
+// InstanceExpirationDay represents the amount of days to expire an instance
+const InstanceExpirationDay = 5
```
Comment (therealplato) [inline services/instance.go]: use `InstanceExpirationDays`, this could be misinterpreted as "thursday" or "friday"

File: instance/instance.go (hunk 1 above)
Comment (cored) [inline instance/instance.go]: @therealplato when the day is 29 it's properly rotated to the next month. I added a new unit test so you can see it working. 
In another note; I didn't see that constant in the time package documentation `Day`. 
https://golang.org/pkg/time/
There's been discussion about adding it but it doesn't exist. 
https://github.com/golang/go/issues/11473

File: instance/instance.go (hunk 2 above)
Comment (cored) [inline instance/instance.go]: @therealplato why do you think that will be better?

File: instance/instance.go (hunk 3 above)
Comment (cored) [inline instance/instance.go]: @therealplato Any reason for going on that route?

File: services/instance.go (hunk 4 above)
Comment (cored) [inline services/instance.go]: sure

File: instance/instance.go (hunk 1 above)
Comment (therealplato) [inline instance/instance.go]: got it, thx

File: instance/instance.go (hunk 3 above)
Comment (therealplato) [inline instance/instance.go]: as usual I dont like adding a new type that does not have behavior attached, and at the moment there is no reason to set GlobalPath dynamically

File: instance/instance.go (hunk 2 above)
Comment (therealplato) [inline instance/instance.go]: slices are already pointers to their elements, i think this makes it hold pointers to pointers

File: instance/instance.go (hunk 3 above)
Comment (cored) [inline instance/instance.go]: @therealplato the code in the service is using this type right now; I can remove it an pass an string if you prefer it like that; however I prefer to change it in another PR.

File: instance/instance.go (hunk 2 above)
Comment (cored) [inline instance/instance.go]: in what way does that create an problem? @therealplato I still can't see the issue.

Comment (therealplato) [inline instance/instance.go]: `[]*Instance` is harder to comprehend and more complicated to use than `[]Instance`

Comment (cored) [inline instance/instance.go]: @therealplato do you mind if we ask for a second opinion? I still can't see what is the complication. @bobbytables @macat @poopoothegorilla

File: instance/instance.go (hunk 1 above)
Comment (bobbytables) [inline instance/instance.go]: A day is a variable unit of time so it probably won't change in Go.

File: instance/instance.go (hunk 3 above)
Comment (bobbytables) [inline instance/instance.go]: I agree with Isaac, this is an over abstraction.

File: instance/instance.go [hunk 5]
```diff
@@ -105,17 +111,12 @@ func FindByPath(store store.Store, path Path) (*Instance, error) {
...
-		}
-		instances = append(instances, instance)
-	}
+	return retrieveInstancesByKey(store, playbookPath.String())
+}
 
-	return instances, nil
+// All find all instances from in the store
+func All(store store.Store, globalPath GlobalPath) ([]*Instance, error) {
+	return retrieveInstancesByKey(store, globalPath.String())
```
Comment (bobbytables) [inline instance/instance.go]: Yeah like the GlobalPath in here is just adding more to the function signature. It's not necessary.

File: instance/instance.go (hunk 3 above)
Comment (cored) [inline instance/instance.go]: @bobbytables Sure; I'll remove this abstraction.

File: server/server.go [hunk 6]
```diff
@@ -74,6 +79,14 @@ func (s *Server) Init() {
 
 	s.playbooks = deployment.AllPlaybooks
 	glog.Infof("Server Playbooks: %+v", s.playbooks)
+
+	glog.Info("Initialize deployed instances cleanup worker")
+	ds := services.NewDeploymentService(s.Cfg, s.store, s.playbooks, s.manifests)
+	go func() {
+		for {
+			t = timer.AfterFunc(time.Hour*24*time.Duration(daysForInstanceCleanup), ds.RemoveExpiredInstances())
```
Comment (poopoothegorilla) [inline server/server.go]: sweet

File: instance/instance.go (hunk 2 above)
Comment (bobbytables) [inline instance/instance.go]: You should really only use a pointer when you need to modify the object when iterating over it. I'm of the fact that we're not going to save valuable memory here by using a pointer. Go uses pointers to show intent as well, and using a pointer here may give mixed signals.

File: services/deployment_test.go [hunk 7]
```diff
@@ -17,6 +18,56 @@ func init() {
...
+	expirationDate := time.Date(2016, 8, 05, 16, 00, 00, 651387237, time.UTC)
+	expiredAt := instance.NewExpiredAt(InstanceExpirationDays, currentTime)
+
+	cases := []struct {
+		Scenario string
+		Instance *instance.Instance
+		Error    error
+	}{
+		{
+			Scenario: "RemoveExpiredInstances: When an instance just expired",
```
Comment (poopoothegorilla) [inline services/deployment_test.go]: are there more scenarios?

File: instance/instance.go (hunk 2 above)
Comment (cored) [inline instance/instance.go]: @bobbytables cool, will remove the pointer declaration.

File: server/server.go (hunk 6 above)
Comment (bobbytables) [inline server/server.go]: Why not do every 15 seconds?

File: server/server.go [hunk 8]
```diff
@@ -28,9 +29,13 @@ type Server struct {
 	Cfg        cfg.Type
 }
 
-const commandHint string = `/broadway help: This message
+const (
+	commandHint string = `/broadway help: This message
 /broadway deploy myPlaybookID myInstanceID: Deploy a new instance`
 
+	daysForInstanceCleanup = 5
```
Comment (poopoothegorilla) [inline server/server.go]: I am guessing this will go away later?

File: services/deployment.go [hunk 9]
```diff
@@ -220,6 +221,26 @@ func (d *DeploymentService) DeleteAndNotify(i *instance.Instance) error {
...
+// RemoveExpiredInstances remove expired instances from the deployment
+func (d *DeploymentService) RemoveExpiredInstances(expirationDate time.Time) error {
+	globalPath := fmt.Sprintf("%s/instances", d.Cfg.EtcdPath)
+	instances, err := instance.All(d.store, globalPath)
+	if err != nil {
+		return err
+	}
+	for _, i := range instances {
+		if i.Status == instance.StatusDeployed {
+			if i.ExpiredAt < expirationDate.Unix() {
```
Comment (bobbytables) [inline services/deployment.go]: nitty gritty but this should be `<=` I think. If it matches exactly it should be considered expired.

File: server/server.go (hunk 6 above)
Comment (cored) [inline server/server.go]: Did not have any time specified so I picked up one. @bobbytables will add the 15 seconds interval.

File: services/instance.go [hunk 10]
```diff
@@ -181,6 +189,9 @@ func (is *InstanceService) Delete(i *instance.Instance) error {
 	return instance.Delete(is.store, path)
 }
 
+// func (is *InstanceService) RemoveExpiredInstances() error {
+// }
```
Comment (bobbytables) [inline services/instance.go]: Ditch?

File: server/server.go (hunk 8 above)
Comment (poopoothegorilla) [inline server/server.go]: or what if you used `instance.InstanceExpirationDays` instead?

File: services/instance.go [hunk 11]
```diff
@@ -181,6 +189,9 @@ func (is *InstanceService) Delete(i *instance.Instance) error {
 	return instance.Delete(is.store, path)
 }
 
+// func (is *InstanceService) RemoveExpiredInstances() error {
```
Comment (poopoothegorilla) [inline services/instance.go]: do we still want this?

Comment (cored) [inline services/instance.go]: @poopoothegorilla good catch, thanks

File: services/deployment.go (hunk 9 above)
Comment (cored) [inline services/deployment.go]: @bobbytables cool

File: store/etcdstore/etcd.go [hunk 12]
```diff
@@ -60,9 +61,12 @@ func (*etcdStore) Values(path string) (values map[string]string) {
 		glog.Warningf("Ignoring error getting values: %s. Error: %s", path, err.Error())
 		return values
 	}
+	fmt.Println(resp)
```
Comment (poopoothegorilla) [inline store/etcdstore/etcd.go]: needed still?

File: server/server.go (hunk 8 above)
Comment (cored) [inline server/server.go]: @poopoothegorilla one variable is to specified the amount of time to create the expiration date the other one is to specified how often we do the cleanup.

File: store/etcdstore/etcd.go [hunk 13]
```diff
@@ -60,9 +61,12 @@ func (*etcdStore) Values(path string) (values map[string]string) {
...
 	if resp.Node != nil && len(resp.Node.Nodes) > 0 {
 		for _, node := range resp.Node.Nodes {
-			values[lastKeyItem(node.Key)] = node.Value
+			if node.Dir == false {
```
Comment (poopoothegorilla) [inline store/etcdstore/etcd.go]: what is this?

File: server/server.go (hunk 8 above)
Comment (poopoothegorilla) [inline server/server.go]: oh good point nice thanks for the clarification

File: instance/instance.go (hunk 2 above)
Comment (cored) [inline instance/instance.go]: @therealplato @bobbytables if you guys checkout the code in `toJSON` function which is the function doing the marshalling is returning a pointer and that was the reason for me to use an slice of instance pointers so I can use append without extracting the value from the pointer. What about if we do this change later, I mean if there's a very specific benefit of changing this right now I can start the refactoring of the other functions in the `instance` package to use a value instead of a pointer. Let me know your opinions.

File: store/etcdstore/etcd.go [hunk 14]
```diff
@@ -62,7 +62,9 @@ func (*etcdStore) Values(path string) (values map[string]string) {
 	}
 	if resp.Node != nil && len(resp.Node.Nodes) > 0 {
 		for _, node := range resp.Node.Nodes {
-			values[lastKeyItem(node.Key)] = node.Value
+			if node.Dir == false {
```
Comment (poopoothegorilla) [inline store/etcdstore/etcd.go]: @cored could you explain what this is doing?

Comment (cored) [inline store/etcdstore/etcd.go]: @poopoothegorilla a node in etcd could be a directory or an actual node. A directory contains several nodes with values but a directory does not contain a value. So we need to make the distinction. I'm writing a test right now to stress this out it's been stressed indirectly by the integration test on the deployment service.

File: services/deployment_test.go (hunk 7 above)
Comment (cored) [inline services/deployment_test.go]: @poopoothegorilla There's a new one for removing if the expiration date is the same as current date.

File: store/etcdstore/etcd.go (hunk 14 above)
Comment (poopoothegorilla) [inline store/etcdstore/etcd.go]: cool just ping me when you got those :)

**Comment by poopoothegorilla** [comment]
when tests pass +1

File: services/deployment.go [hunk 15]
```diff
@@ -220,6 +221,26 @@ func (d *DeploymentService) DeleteAndNotify(i *instance.Instance) error {
 
 }
 
+// RemoveExpiredInstances remove expired instances from the deployment
+func (d *DeploymentService) RemoveExpiredInstances(expirationDate time.Time) error {
+	globalPath := fmt.Sprintf("%s/instances", d.Cfg.EtcdPath)
```
Comment (macat) [inline services/deployment.go]: Can you add logging here. I would like to see when this method runs and also some info about how many instances it found.

File: server/server.go [hunk 16]
```diff
@@ -74,6 +72,14 @@ func (s *Server) Init() {
 
 	s.playbooks = deployment.AllPlaybooks
 	glog.Infof("Server Playbooks: %+v", s.playbooks)
+
+	glog.Info("Initialize deployed instances cleanup worker")
+	ds := services.NewDeploymentService(s.Cfg, s.store, s.playbooks, s.manifests)
+	go func() {
+		for {
+			t = timer.AfterFunc(time.Hour*12*time.Duration(time.Hour), ds.RemoveExpiredInstances())
```
Comment (macat) [inline server/server.go]: This needs to be configurable from env vars.

File: instance/instance.go (hunk 1 above)
Comment (therealplato) [inline instance/instance.go]: this actually shouldn't work because you're not assigning `from` before you return it
i still suggest `return from.Add(time.Hour * 24 * time.Duration(daysToExpire))`

File: instance/instance.go [hunk 17]
```diff
@@ -72,24 +80,15 @@ const (
 	// StatusNew represents a newly created instance
 	StatusNew Status = ""
 	// StatusDeploying represents an instance that has begun deployment
-	StatusDeploying = "deploying"
+	StatusDeploying Status = "deploying"
```
Comment (therealplato) [inline instance/instance.go]: why did you add these?

File: instance/instance.go [hunk 18]
```diff
@@ -105,17 +104,12 @@ func FindByPath(store store.Store, path Path) (*Instance, error) {
 
 // FindByPlaybookID find all the instances for an specified playbook path
```
Comment (therealplato) [inline instance/instance.go]: let's rename this if it's no longer taking just the playbook ID

File: instance/instance.go [hunk 19]
```diff
@@ -177,3 +171,17 @@ func toJSON(instance *Instance) (string, error) {
 	}
 	return string(encoded), nil
 }
+
+func retrieveInstancesByKey(store store.Store, key string) ([]*Instance, error) {
```
Comment (therealplato) [inline instance/instance.go]: we can move this behavior into All and remove this function

File: instance/instance.go (hunk 17 above)
Comment (cored) [inline instance/instance.go]: @therealplato an string is not the same as an `Status` when I was asserting on the status the type check failed.

File: instance/instance.go (hunk 18 above)
Comment (cored) [inline instance/instance.go]: @therealplato ? I don't get what you mean.

File: instance/instance.go (hunk 19 above)
Comment (cored) [inline instance/instance.go]: ok

File: instance/instance_test.go [hunk 20]
```diff
@@ -329,3 +369,35 @@ func TestUnlock(t *testing.T) {
...
+		ExpectedInstances []*Instance
+		ExpectedError     error
+	}{
+		{
+			Scenario: "All: When the instance is properly save",
+			Path:     "broadwaytest/instances",
+			Store: &store.FakeStore{
+				MockValues: func(path string) map[string]string {
+					return map[string]string{
+						"etcdPath/instances": `{"playbook_id":"test", "id": "id", "status": "deployed"}`,
```
Comment (therealplato) [inline instance/instance_test.go]: can you return more than one instance here? in real life usage we will usually have more than one and i'd like the test to reflect that scenario

File: server/server.go [hunk 21]
```diff
@@ -28,9 +29,6 @@ type Server struct {
 	Cfg        cfg.Type
 }
 
-const commandHint string = `/broadway help: This message
-/broadway deploy myPlaybookID myInstanceID: Deploy a new instance`
-
```
Comment (therealplato) [inline server/server.go]: thx

File: services/deployment.go [hunk 22]
```diff
@@ -220,6 +221,26 @@ func (d *DeploymentService) DeleteAndNotify(i *instance.Instance) error {
...
+	instances, err := instance.All(d.store, globalPath)
+	if err != nil {
+		return err
+	}
+	for _, i := range instances {
+		if i.Status == instance.StatusDeployed {
```
Comment (therealplato) [inline services/deployment.go]: should we remove deployed but locked instances @bobbytables @macat

File: instance/instance_test.go (hunk 20 above)
Comment (cored) [inline instance/instance_test.go]: @therealplato there is a test reflecting that in the `etcd_test`. What will be the benefit of also adding it in here?

File: services/deployment.go [hunk 23]
```diff
@@ -220,6 +221,26 @@ func (d *DeploymentService) DeleteAndNotify(i *instance.Instance) error {
...
+			if i.ExpiredAt <= expirationDate.Unix() {
```
Comment (therealplato) [inline services/deployment.go]: let's cache `x := expirationDate.Unix()` outside this loop

File: services/instance.go [hunk 24]
```diff
@@ -15,6 +15,9 @@ import (
 	"github.com/namely/broadway/store"
 )
 
+// InstanceExpirationDays represents the amount of days to expire an instance
+const InstanceExpirationDays = 5
```
Comment (therealplato) [inline services/instance.go]: can you add `TODO: inject via config objects`

File: instance/instance.go (hunk 17 above)
Comment (therealplato) [inline instance/instance.go]: @macat we were wrong about this behavior :o

File: instance/instance.go (hunk 18 above)
Comment (therealplato) [inline instance/instance.go]: gorename to "FindByPlaybookPath" because the path is not the id

File: instance/instance_test.go (hunk 20 above)
Comment (therealplato) [inline instance/instance_test.go]: you would be testing "the All function successfully converts multiple etcd json into instances", the previous test is only testing retrieving json

Comment (cored) [inline instance/instance_test.go]: I will remove this function either way

File: instance/instance.go (hunk 17 above)
Comment (macat) [inline instance/instance.go]: ups, sorry.

**Comment by macat** [comment]
+1
